.circleci/config.yml

```diff
@@ -1,141 +1,66 @@
-# Java Maven CircleCI 2.0 configuration file
-#
-# Check https://circleci.com/docs/2.0/language-java/ for more details
-#
-version: 2.0
-
-#########################
-# CACHES
-#########################
-
-restore_cache: &restore_m2_cache
-  keys:
-    - m2-{{ checksum "pom.xml" }}
-    - m2- # used if checksum fails
-
-#########################
-# COMMANDS
-#########################
-
-commands:
-
-  restore_m2_cache:
-    description: "Restore m2 cache for specified maven task"
-    parameters:
-      cache_name:
-        default: ""
-        type: enum
-        enum: ["", "security-proxy-request-mapper"]
-    steps:
-      - restore_cache:
-          keys:
-            - m2-<<parameters.cache_name>>-{{ checksum "pom.xml" }}
-            - m2-<<parameters.cache_name>>- # used if checksum fails
-            - m2-
-
-  save_m2_cache:
-    description: "Save m2 cache for specified maven task"
-    parameters:
-      cache_name:
-        default: ""
-        type: enum
-        enum: ["", "security-proxy-request-mapper"]
-    steps:
-      - save_cache:
-          key: m2-<<parameters.cache_name>>-{{ checksum "pom.xml" }}
-          paths:
-            - ~/.m2
+version: 2
 
 jobs:
-
   #########################
   # BUILD
   #########################
   build:
+    working_directory: ~/repo
     docker:
       - image: circleci/openjdk:8-jdk
-    working_directory: ~/repo
+        environment:
+          MAVEN_CLI_OPTS: --threads 1C --batch-mode --errors --fail-at-end --show-version -DinstallAtEnd=true -DdeployAtEnd=true -s .circleci/mavenSettings.xml
+          MAVEN_OPTS: -Dorg.slf4j.simpleLogger.log.org.apache.maven.cli.transfer.Slf4jMavenTransferListener=WARN -Dorg.slf4j.simpleLogger.showDateTime=true
+
     steps:
-      - setup_remote_docker:
-          docker_layer_caching: true
       - checkout
-      - restore_cache: *restore_m2_cache
+      - restore_cache:
+          key: example-security-proxy-request-mapper-{{ checksum "pom.xml" }}
       - run:
-          name: Build and Unit test
-          command: mvn -s .circleci/mavenSettings.xml clean install -U
+          name: Collect and save dependencies
+          command: mvn $MAVEN_CLI_OPTS dependency:go-offline
       - save_cache:
-          key: m2-{{ checksum "pom.xml" }}
           paths:
             - ~/.m2
-      - persist_to_workspace:
-          root: ~/
-          paths:
-            - repo
+          key: example-security-proxy-request-mapper-{{ checksum "pom.xml" }}
+      - run:
+          name: Build and run unit tests
+          command: mvn $MAVEN_CLI_OPTS package
```

Comment:
You're not storing any artifacts as a result of this build step. I would expect that we at least store the jars (and maybe also the test report).

See also a next comment we should use these jars again in the build docker step. 

There are a few very important things to note down when creating a build pipeline:
- Speed: this means that you want to have the least possible amount of duplication (with Maven that is a bit harder then is should be though).
- Reproducability: Every run of the pipeline must always produce exactly the same result regardless whether the pipeline is run at a different moment in time. This means that you should not depend on any external things such as an external artifact repository (Artifactory). These may cause unexpected behaviour in your pipeline when the pipeline is run at different moments in time.

Reply:
This project doesn't contains any tests, I thought jar built in this step will be used in dockers phase because in workflows there is a definition which tells that build step is required to run dockers stuff

Reply:
No, it's not magically done since you haven't defined any output of the build step 😄.

It's about this part of the language guide:
```
      - store_artifacts: # store the uberjar as an artifact
      # Upload test summary for display in Artifacts: https://circleci.com/docs/2.0/artifacts/
          path: target/demo-java-spring-0.0.1-SNAPSHOT.jar
      # See https://circleci.com/docs/2.0/deployment-integrations/ for deploy examples
```

Reply:
but on the bottom in workflows, there is a task responsible for deploying artifacts, I decided to use maven because we use it to build app, dockers etc. so maybe it is a good idea to stick with it because everyone know what is happening after mvn deploy, what do you think? :D

Reply:
As mentioned during our chat:
- Please don't upload the maven artifacts to artifactory for this project as the main build artifact of this project is a Docker container. 
- Please store the build artifacts (jars) using `store_artifacts` so we can reuse them in the build docker steps.